deckhouse-controller/pkg/controller/deckhouse-release/controller.go

```diff
@@ -0,0 +1,455 @@
+	}
+	imageHash := imageID[idSplitIndex+1:]
+
+	image := pods[0].Spec.Containers[0].Image
+	imageRepoTag, err := gcr.NewTag(image)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("incorrect image: %s", image)
+	}
+
+	repo := imageRepoTag.Context().Name()
+	tag := imageRepoTag.TagStr()
+
+	registrySecret, err := r.getRegistrySecret(ctx)
+	if apierrors.IsNotFound(err) {
+		err = nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var opts []cr.Option
+	if registrySecret != nil {
+		opts = []cr.Option{
+			cr.WithCA(string(registrySecret.Data["ca"])),
+			cr.WithInsecureSchema(string(registrySecret.Data["scheme"]) == "http"),
+			cr.WithAuth(string(registrySecret.Data[".dockerconfigjson"])),
+		}
+	}
+
+	regClient, err := r.dc.GetRegistryClient(repo, opts...)
+	if err != nil {
+		r.logger.Errorf("Registry (%s) client init failed: %s", repo, err)
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	repoDigest, err := regClient.Digest(tag)
+	if err != nil {
+		r.logger.Errorf("Registry (%s) get digest failed: %s", repo, err)
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	if strings.TrimSpace(repoDigest) == strings.TrimSpace(imageHash) {
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	r.logger.Info("New deckhouse image found. Restarting")
+	now := time.Now().Format(time.RFC3339)
+	if os.Getenv("D8_IS_TESTS_ENVIRONMENT") != "" {
+		now = time.Date(2021, 01, 01, 13, 30, 00, 00, time.UTC).Format(time.RFC3339)
+	}
+
+	annotationsPatch := map[string]interface{}{
+		"spec": map[string]interface{}{
+			"template": map[string]interface{}{
+				"metadata": map[string]interface{}{
+					"annotations": map[string]string{
+						"kubectl.kubernetes.io/restartedAt": now,
+					},
+				},
+			},
+		},
+	}
+
+	jsonPatch, _ := json.Marshal(annotationsPatch)
+	err = r.client.Patch(
+		ctx,
+		&appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: pods[0].Namespace,
+				Name:      "deckhouse",
+			},
+		},
+		client.RawPatch(types.MergePatchType, jsonPatch),
+	)
+
+	if err != nil {
+		r.logger.Errorf("Patch deckhouse deploymaent failed: %s", err)
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+}
+
+func (r *deckhouseReleaseReconciler) getRegistrySecret(ctx context.Context) (*corev1.Secret, error) {
+	var secret corev1.Secret
+	key := types.NamespacedName{Namespace: "d8-system", Name: "deckhouse-registry"}
+	err := r.client.Get(ctx, key, &secret)
+	if err != nil {
+		return nil, fmt.Errorf("get secret %s: %w", key, err)
+	}
+
+	return &secret, nil
+}
+
+// TODO: get from release object
+func (r *deckhouseReleaseReconciler) getReleaseData(ctx context.Context) (updater.DeckhouseReleaseData, error) {
+	var cm corev1.ConfigMap
+
+	key := types.NamespacedName{Namespace: "d8-system", Name: "d8-release-data"}
+	err := r.client.Get(ctx, key, &cm)
```

Comment:
Also have to add [here](https://github.com/deckhouse/deckhouse/blob/main/deckhouse-controller/pkg/controller/controller.go#L131)

Reply:
Done